services/QuillLMS/db/migrate/20251029210337_convert_examples_to_sti.evidence.rb

```diff
@@ -0,0 +1,370 @@
+    return [] unless ActiveRecord::Base.connection.column_exists?(table_name, fk_column)
+
+    count = check_foreign_key_count(table_name, fk_column, referenced_table)
+    return [] if count.zero?
+
+    ["#{table_name}.#{fk_column} (#{count} rows)"]
+  end
+
+  private def check_foreign_key_count(referencing_table, fk_column, referenced_table)
+    result = execute(<<-SQL.squish)
+      SELECT COUNT(*)
+      FROM #{referencing_table}
+      WHERE #{fk_column} IN (SELECT id FROM #{referenced_table})
+    SQL
+    result.getvalue(0, 0).to_i
+  end
+
+  private def raise_if_unexpected_references_found(unexpected_references)
+    return if unexpected_references.empty?
+
+    error_msg = "Found unexpected foreign key references that need to be migrated:\n  - #{unexpected_references.join("\n  - ")}"
+    Rails.logger.error error_msg
+    raise error_msg
+  end
+
+  private def migrate_prompt_examples
+    count = 0
+    PromptExample.without_new_example.find_each do |prompt_example|
+      Example.create!(
+        prompt_example.attributes.except('id').merge(
+          type: 'Evidence::Research::GenAi::PromptExample',
+          original_type: 'PromptExample',
+          original_id: prompt_example.id
+        )
+      )
+      count += 1
+    end
+    Rails.logger.debug "Migrated #{count} PromptExamples"
+  end
+
+  private def migrate_test_examples
+    count = 0
+    TestExample.without_new_example.find_each do |test_example|
+      Example.create!(
+        test_example.attributes.except('id').merge(
+          type: 'Evidence::Research::GenAi::TestExample',
+          original_type: 'TestExample',
+          original_id: test_example.id
+        )
+      )
+      count += 1
+    end
+    Rails.logger.debug "Migrated #{count} TestExamples"
+  end
+
+  private def migrate_draft_examples
+    count = 0
+    DraftExample.without_new_example.find_each do |draft_example|
+      Example.create!(
+        draft_example.attributes.except('id').merge(
+          type: 'Evidence::Research::GenAi::DraftExample',
+          original_type: 'DraftExample',
+          original_id: draft_example.id
+        )
+      )
+      count += 1
+    end
+    Rails.logger.debug "Migrated #{count} DraftExamples"
+  end
+
+  private def reset_sequence
+    return unless Example.any?
+
+    ActiveRecord::Base.connection.reset_pk_sequence!('evidence_research_gen_ai_examples')
+    Rails.logger.debug 'Reset primary key sequence for evidence_research_gen_ai_examples'
+  end
+
+  private def update_foreign_keys
+    Rails.logger.debug 'Updating foreign keys...'
+
+    prompt_fk_count = update_prompt_example_foreign_keys
+    llm_example_fk_count = update_test_example_foreign_keys
+
+    Rails.logger.debug "Updated #{prompt_fk_count} llm_prompt_prompt_examples, #{llm_example_fk_count} llm_examples"
+  end
+
+  private def update_prompt_example_foreign_keys
+    referenced_ids = LlmPromptPromptExample.distinct.pluck(:prompt_example_id).compact
```

Comment:
Won't this be like 200K ids put into a SQL statement? Do we need this? Couldn't the update just do every one (if the record was already updated, it just wouldn't change anything)

Reply:
Doing it one-by-one was one of the bottlenecks I encountered.  This ended up being pretty fast on staging.